api.visual/src/org/netbeans/api/visual/model/ObjectScene.java

```diff
@@ -86,16 +86,62 @@ public final void addObject (Object object, Widget... widgets) {
         objects.put (object, object);
         object2widget.put (object, mainWidget);
         object2widgets.put (object, mainWidget != null ? Arrays.asList (widgets) : EMPTY_WIDGETS_LIST);
-        objectStates.put (object, ObjectState.createNormal ());
-
+        ObjectState state = objectStates.computeIfAbsent(object, (o) -> ObjectState.createNormal());
         for (Widget widget : widgets) {
             widget2object.put (widget, object);
-            widget.setState (ObjectState.createNormal ());
+            widget.setState (state);
         }
 
         for (ObjectSceneListener listener : getListeners (ObjectSceneEventType.OBJECT_ADDED))
             listener.objectAdded (event, object);
     }
+    
+    /**
+     * Removes mapping for an object, including the object state. The caller is responsible for removing
+     * widgets from the scene before this call. If the is already removed or did not exist, the method
+     * does nothing. The method <b>will not</b> clear object's state.
+     * 
+     * @param object object, whose mapping should be removed
```

Comment:
Out of curiosity and without knowing visual library much, I'd welcome a pointer to function/constructor that creates/registers the object that can be passed here as an argument.

Reply:
Ouch! I need to correct javadoc! Because the method actually does **not** clear the state (intentionally). To answer your question - an ordinary call to addObject(Widget...) registers an object (and binds widgets to it). This removeMapping() just removes mapping to the widgets from maps, so widgets could be eventually GCed. 
The usage, for an implementation that uses the lazy- features, is that widgets, which are outside any viewport can be removed from the scene (saving tree processing) and can be removed from the object-widget mapping here. But, their potential selected/highlighted state should be left as it is. They have just no widgets at the moment.